batch/sql/estimated-current.sql

```diff
@@ -1205,18 +1215,24 @@ BEGIN
     SET state = new_state, status = new_status, attempt_id = in_attempt_id
     WHERE batch_id = in_batch_id AND job_id = in_job_id;
 
-    UPDATE batches SET n_completed = n_completed + 1 WHERE id = in_batch_id;
-    UPDATE batches
+    UPDATE batches_n_jobs_in_complete_states
+      SET n_completed = n_completed + 1,
+          n_cancelled = n_cancelled + (new_state = 'Cancelled'),
+          n_failed    = n_failed + (new_state = 'Error' OR new_state = 'Failed'),
+          n_succeeded = n_succeeded + (new_state != 'Cancelled' AND new_state != 'Error' AND new_state != 'Failed')
+      WHERE id = in_batch_id;
+
+    SELECT n_completed INTO new_n_completed
+    FROM batches_n_jobs_in_complete_states
+    WHERE id = in_batch_id;
+
+    # Grabbing an exclusive lock on batches here could deadlock,
+    # but this IF should only execute for the last job
+    IF new_n_completed = total_jobs_in_batch THEN
+      UPDATE batches
       SET time_completed = new_timestamp,
           `state` = 'complete'
-      WHERE id = in_batch_id AND n_completed = batches.n_jobs;
-
-    IF new_state = 'Cancelled' THEN
-      UPDATE batches SET n_cancelled = n_cancelled + 1 WHERE id = in_batch_id;
-    ELSEIF new_state = 'Error' OR new_state = 'Failed' THEN
-      UPDATE batches SET n_failed = n_failed + 1 WHERE id = in_batch_id;
-    ELSE
-      UPDATE batches SET n_succeeded = n_succeeded + 1 WHERE id = in_batch_id;
+      WHERE id = in_batch_id;
```

Comment:
You don't need to fix here b/c it's already a present bug, but something I just noticed that you should keep in mind going forward: I think time_completed can jitter. It really should only change from NULL to a timestamp *once*.

1. T1 does n++
2. T2 does n++
3. T1 reads n_jobs == n_completed
4. T2 reads n_jobs == n_completed
5. T1 updates with its new_timestamp
6. T2 updates with its new_timestamp

I'm not sure how to solve it the "right" way without a `max` over all jobs attempts.

Reply:
Would 3 be true? I thought the point of repeatable read would be that T1 should always be able to select after an update and see its own write. It should see some stale view of the table and only the last job to grab the X lock on n_completed will actually see n_jobs == n_completed

Reply:
Or actually even stronger, the procedure gets serialized starting at the first `UPDATE`, since it must hold an exclusive lock through the end of the transaction (in case it needs to be rolled back) and then your order of 1-6 wouldn't actually be interleaved in that way

Reply:
T1 is able to see its own write, but might also see subsequent writes, which it does in my example. There's a really clear example of this here: https://dev.mysql.com/doc/refman/5.7/en/innodb-consistent-read.html.